### PR TITLE
Migrate to TypeScript, Fix audio Safari support and NodeJS Suppport

### PR DESCRIPTION
**Pull request by hthetiot**
# ChangesLogs
1. Migrate to TypeScript
2. Fix Safari Audio  by implemeting createConstantSource fallback
3. Keep video ratio when rendering
4. Fix NodeJS support
5. Move canvas creation into start()
 
# PR Demo URL: 
- https://hthetiot.github.io/video-stream-merger/

# Tested:
- Chrome: OK
- MAcOS Safari Version 14.1 (16611.1.21.161.3):  OK (fixed compare to original version)
- iOS 14.4, 14.5, 15.6: OK (fixed compare to original version, but still get black stream under most conditions due ios bug)
- Firefox 89.0.2: FAIL (same on original version)

iOS related issue for "Video Element cannot playback local Canvas.captureStream on iOS" fixed on iOS 15: - https://bugs.webkit.org/show_bug.cgi?id=181663

### CODE REVIEW

**Comment by hthetiot** [comment]
@t-mullen Hello, I'm using TypeScript and needed some fixes on Safari/Rendering.
Let me know if you have any review feedback.

**Comment by hthetiot** [comment]
Let me know @t-mullen if you want me to make seprate PR for iOS support (createConstantSource/canvas on start instead of construtor).

File: src/index.ts [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+        height: canvasSize.height / size.height,
+      };
+
+      const ratio  = Math.min ( sizeRatio.height, sizeRatio.width);
+
+      position.x = ( canvasSize.width - size.width * ratio ) / 2;
+      position.y = ( canvasSize.height - size.height * ratio ) / 2;
+
+      try {
+        this._ctx?.drawImage(element, 0, 0, size.width, size.height, position.x, position.y, size.width*ratio, size.height*ratio);
```
Comment (t-mullen) [inline src/index.ts]: Why force (0,0) when aspect ratio is being enforced?

File: src/index.ts [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+    }
+  }
+
+  _drawVideo(element: HTMLVideoElement, stream: any) {
+
+
+    // default draw function
+    const canvasSize = { height: this.height, width: this.width};
+
+    const position = {
```
Comment (t-mullen) [inline src/index.ts]: Would prefer not to allocate an object in the draw loop, unless benchmarks show it doesn't matter.

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+      x: stream.x || 0,
+      y: stream.y || 0
+    };
+
+    const size = {
+        height: stream.height || element.videoHeight || canvasSize.height,
```
Comment (t-mullen) [inline src/index.ts]: Using the videoWidth over canvas.width is new behaviour - is this intentional?

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+
+    // TODO move to sreeam option to enable new behavior
+    const keepRatio = false;
+
+    if (!keepRatio) {
+
+      try {
+          this._ctx?.drawImage(element, position.x, position.y, size.width, size.height);
+      } catch {
+        // Ignore error
```
Comment (t-mullen) [inline src/index.ts]: We should log this error even if we can't handle it.

**Review by t-mullen — CHANGES_REQUESTED** [review]
Looks great! Thanks for going through the effort here.

A few thoughts on the changes to the draw loop, everything else looks good.

File: src/index.ts (hunk 1 above)
Comment (t-mullen) [reply src/index.ts]: Nevermind - I didn't notice this is the sx,sy version of drawImage.

File: src/index.ts [hunk 5]
```diff
@@ -0,0 +1,508 @@
...
+      y: stream.y || 0
+    };
+
+    const size = {
+        height: stream.height || element.videoHeight || canvasSize.height,
+        width: stream.width || element.videoWidth || canvasSize.width
+    };
+
+    // TODO move to sreeam option to enable new behavior
+    const keepRatio = false;
```
Comment (t-mullen) [inline src/index.ts]: I like this feature, moving it into an option is a little awkward though.

Maybe - as the default behaviour - we should maintain the aspect ratio if a width or height isn't specified (and fill the canvas if neither are).

File: src/index.ts (hunk 2 above)
Comment (hthetiot) [reply src/index.ts]: Fixed by c068b34

File: src/index.ts (hunk 4 above)
Comment (hthetiot) [reply src/index.ts]: Fixed by c068b34

File: src/index.ts (hunk 3 above)
Comment (hthetiot) [reply src/index.ts]: This is in order to get the actual video size in order to keep the ratio.
We may only do that if keepRatio is true, if you d'like.

File: src/index.ts (hunk 5 above)
Comment (hthetiot) [reply src/index.ts]: I think option may be needed to prevent unexpected new behaviour.
Let me know and i will make changes as you want it to be.

**Comment by hthetiot** [comment]
@t-mullen ready for more review.
Notice that I incremented version to 4.0.0 in package.json because i think it may allow you to maintain current 3.X separately if needed.

**Comment by hthetiot** [comment]
Next i can make PR for merging WebGL implementation, then another for offScreen canvas.

File: src/index.ts (hunk 5 above)
Comment (t-mullen) [reply src/index.ts]: New behaviour is fine in a major version (I doubt many applications are using a warped aspect ratio anyways).

That said - If you can remove this feature from this PR, I'll open another one to implement fixed aspect ratio. Don't want to block the rest of your fixes. 👍

File: src/index.ts (hunk 3 above)
Comment (t-mullen) [reply src/index.ts]: I agree aspect ratio should be kept, but not specifying a width and height should still fullscreen the stream.

Revert this for now - will address in another PR.

File: src/index.ts (hunk 5 above)
Comment (hthetiot) [reply src/index.ts]: Ok I will revert this part.
Thank you.

**Comment by hthetiot** [comment]
Ready for review @t-mullen

**Comment by t-mullen** [comment]
Published as 4.0.0. Thanks again!

**Comment by hthetiot** [comment]
Thank to you.
Will make PR for keep video ratio and webGL branch merge with adapter pattern next.

**Pull request merged**